scio-core/src/main/scala/com/spotify/scio/coders/CoderMaterializer.scala

```diff
@@ -17,20 +17,91 @@
 
 package com.spotify.scio.coders
 
-import org.apache.beam.sdk.coders.{Coder => BCoder, IterableCoder, KvCoder, NullableCoder}
+import com.spotify.scio.util.RemoteFileUtil
+import org.apache.beam.sdk.coders.{
+  Coder => BCoder,
+  IterableCoder,
+  KvCoder,
+  NullableCoder,
+  ZstdCoder
+}
 import org.apache.beam.sdk.options.{PipelineOptions, PipelineOptionsFactory}
 
+import java.net.URI
+import java.nio.file.Files
+import java.util.concurrent.ConcurrentHashMap
 import scala.collection.concurrent.TrieMap
+import scala.jdk.CollectionConverters._
 import scala.util.chaining._
 
 object CoderMaterializer {
   import com.spotify.scio.ScioContext
 
-  private[scio] case class CoderOptions(nullableCoders: Boolean, kryo: KryoOptions)
+  private[scio] case class CoderOptions(
+    nullableCoders: Boolean,
+    kryo: KryoOptions,
+    zstdDictMapping: Map[String, Array[Byte]]
+  )
   private[scio] object CoderOptions {
+    private val cache: ConcurrentHashMap[PipelineOptions, CoderOptions] = new ConcurrentHashMap()
+
     final def apply(o: PipelineOptions): CoderOptions = {
-      val nullableCoder = o.as(classOf[com.spotify.scio.options.ScioOptions]).getNullableCoders
-      new CoderOptions(nullableCoder, KryoOptions(o))
+      cache.computeIfAbsent(
+        o,
+        { o =>
+          val scioOpts = o.as(classOf[com.spotify.scio.options.ScioOptions])
+          val nullableCoder = scioOpts.getNullableCoders
+          val zstdDictPaths = Option(scioOpts.getZstdDictionary)
+            .map(_.asScala)
+            .getOrElse(List.empty)
+            .map { s =>
+              // worst api?
+              s.split(":", 2).toList match {
+                case className :: path :: Nil =>
+                  try {
+                    // ensure class exists
+                    Class.forName(className)
+                  } catch {
+                    case e: ClassNotFoundException =>
+                      throw new IllegalArgumentException(
+                        s"Class for zstdDictionary argument ${s} not found.",
+                        e
+                      )
+                  }
+                  className.replaceAll("\\$", ".") -> path
```

Comment:
Maybe missing something cases here? `Class.forName` requires `OuterClass$InnerClass` but the typename in the coders is `OuterClass.InnerClass`

Reply:
A blacklist based on package name may also make sense

Reply:
Added blacklist